lib/Frontend/DiagnosticVerifier.cpp

```diff
@@ -237,68 +278,68 @@ verifyUnknown(SourceManager &SM,
 /// diagnostic \p D.
 bool DiagnosticVerifier::checkForFixIt(const ExpectedFixIt &Expected,
                                        const CapturedDiagnosticInfo &D,
-                                       StringRef buffer) {
+                                       unsigned BufferID) const {
   for (auto &ActualFixIt : D.FixIts) {
     if (ActualFixIt.getText() != Expected.Text)
       continue;
 
-    CharSourceRange Range = ActualFixIt.getRange();
-    if (getColumnNumber(buffer, getRawLoc(Range.getStart())) !=
-        Expected.StartCol)
+    FixItRange ActualRange = ActualFixIt.getComputedRange(
+        SM, BufferID,
+        // Don't compute line numbers unless we have to.
+        /*ComputeStartLocLine=*/Expected.Range.StartLine != FixItRange::NoValue,
+        /*ComputeEndLocLine=*/Expected.Range.EndLine != FixItRange::NoValue);
+
+    if (Expected.Range.StartCol != ActualRange.StartCol ||
+        Expected.Range.EndCol != ActualRange.EndCol) {
+      continue;
+    }
+    if (Expected.Range.StartLine != FixItRange::NoValue &&
+        Expected.Range.StartLine != ActualRange.StartLine) {
       continue;
-    if (getColumnNumber(buffer, getRawLoc(Range.getEnd())) != Expected.EndCol)
+    }
+    if (Expected.Range.EndLine != FixItRange::NoValue &&
+        Expected.Range.EndLine != ActualRange.EndLine) {
       continue;
-
+    }
     return true;
   }
 
   return false;
 }
 
-std::string
-DiagnosticVerifier::renderFixits(ArrayRef<DiagnosticInfo::FixIt> fixits,
-                                 StringRef InputFile) {
+std::string DiagnosticVerifier::renderFixits(ArrayRef<CapturedFixItInfo> fixits,
+                                             unsigned BufferID) const {
   std::string Result;
   llvm::raw_string_ostream OS(Result);
-  interleave(fixits,
-             [&](const DiagnosticInfo::FixIt &ActualFixIt) {
-               CharSourceRange Range = ActualFixIt.getRange();
-
-               OS << "{{"
-                  << getColumnNumber(InputFile, getRawLoc(Range.getStart()))
-                  << '-'
-                  << getColumnNumber(InputFile, getRawLoc(Range.getEnd()))
-                  << '=';
-
-               for (auto C : ActualFixIt.getText()) {
-                 if (C == '\n')
-                   OS << "\\n";
-                 else if (C == '}' || C == '\\')
-                   OS << '\\' << C;
-                 else
-                   OS << C;
-               }
-               OS << "}}";
-             },
-             [&] { OS << ' '; });
+  interleave(
+      fixits,
+      [&](const CapturedFixItInfo &ActualFixIt) {
+        FixItRange ActualRange =
+            ActualFixIt.getComputedRange(SM, BufferID,
+                                         /*ComputeStartLocLine=*/true,
+                                         /*ComputeEndLocLine=*/true);
+
+        OS << "{{"
+           << ActualRange.StartLine << ':' << ActualRange.StartCol
+           << '-'
+           << ActualRange.EndLine << ':' << ActualRange.EndCol
```

Comment:
I am not sure if there's a reliable way to associate an actual fix-it with an expected one, given that their respective order, quantity and contents may differ. It seems like we can at best search for an expected fix-it with matching contents to make this happen. Do you have anything in mind?

Reply:
I just saying something like this:

```cpp
std::string DiagnosticVerifier::renderFixits(ArrayRef<CapturedFixItInfo> fixits,
                                             unsigned BufferID,
                                             unsigned DiagnosticLine) {
 
...

        OS << "{{";

        if (ActualRange.StartLine != DiagnosticLine)
          OS << ActualRange.StartLine << ':';
        OS << ActualRange.StartCol;

        OS << '-';

        if (ActualRange.EndLine != ActualRange.StartLine)
          OS << ActualRange.EndLine << ':';
        OS << ':' << ActualRange.EndCol;

        OS << '=';

        ...
}
```
For example, for:
```
...    // expected-error {{...}} {{10-13=foobar}}
```
But if the actual fix-it were from column 10 to 14 on the same line, we should suggest `{{10-14=foobar}}` instead of something like `{{24:10-24:14=foobar}}` because it's just noisy (, and redundant if we do "implicit same line check" as I suggested in another comment).

Am I missing something?

Reply:
Ah, I misunderstood. SGTM